control-plane/cni/main.go

```diff
@@ -0,0 +1,223 @@
+	// CNIBinDir is the location of the cni config files on the node. Can bet as a cli flag.
+	CNIBinDir string `json:"cni_bin_dir"`
+	// CNINetDir is the locaion of the cni plugin on the node. Can be set as a cli flag.
+	CNINetDir string `json:"cni_net_dir"`
+	// Multus is if the plugin is a multus plugin. Can be set as a cli flag.
+	Multus bool `json:"multus"`
+	// Kubeconfig file name. Can be set as a cli flag.
+	Kubeconfig string `json:"kubeconfig"`
+	// LogLevl is the logging level. Can be set as a cli flag.
+	LogLevel string `json:"log_level"`
+}
+
+// parseConfig parses the supplied configuration (and prevResult) from stdin.
+func parseConfig(stdin []byte) (*PluginConf, error) {
+	cfg := PluginConf{}
+
+	if err := json.Unmarshal(stdin, &cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse network configuration: %v", err)
+	}
+
+	// Parse previous result. This will parse, validate, and place the
+	// previous result object into conf.PrevResult. If you need to modify
+	// or inspect the PrevResult you will need to convert it to a concrete
+	// versioned Result struct.
+	if err := version.ParsePrevResult(&cfg.NetConf); err != nil {
+		return nil, fmt.Errorf("could not parse prevResult: %v", err)
+	}
+	// End previous result parsing
+
+	// Do any validation here
+	// TODO: Do validation
+	//	if conf.AnotherAwesomeArg == "" {
+	//		return nil, fmt.Errorf("anotherAwesomeArg must be specified")
+	//	}
+
+	return &cfg, nil
+}
+
+// cmdAdd is called for ADD requests.
+func cmdAdd(args *skel.CmdArgs) error {
+	cfg, err := parseConfig(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	// Get the values of args passed through CNI_ARGS
+	cniArgs := CNIArgs{}
+	if err := types.LoadArgs(args.Args, &cniArgs); err != nil {
+		return err
+	}
+
+	podNamespace := string(cniArgs.K8S_POD_NAMESPACE)
+	podName := string(cniArgs.K8S_POD_NAME)
+
+	// We only run in a pod
+	if podNamespace == "" && podName == "" {
+		return fmt.Errorf("not running in a pod, namespace and pod should have values")
+	}
+
+	logPrefix := fmt.Sprintf("%s/%s", podNamespace, podName)
+	logger := hclog.New(&hclog.LoggerOptions{
+		Name:  logPrefix,
+		Level: hclog.LevelFromString(cfg.LogLevel),
+	})
+
+	// Check to see if the plugin is a chained plugin
+	if cfg.PrevResult == nil {
+		return fmt.Errorf("must be called as chained plugin")
+	}
+
+	logger.Debug("consul-cni plugin config", "config", cfg)
+	// Convert the PrevResult to a concrete Result type that can be modified. The CNI standard says
+	// that the previous result needs to be passed onto the next plugin
+	prevResult, err := current.GetResult(cfg.PrevResult)
+	if err != nil {
+		return fmt.Errorf("failed to convert prevResult: %v", err)
+	}
+
+	if len(prevResult.IPs) == 0 {
+		return fmt.Errorf("got no container IPs")
+	}
+
+	// Pass the prevResult through this plugin to the next one
+	result := prevResult
+	logger.Debug("consul-cni previous result", "result", result)
+
+	ctx := context.Background()
```

Comment:
In the installer we are generating a kubeconfig file. What is here right now is to prove out that we can generate a proper kubeconfig file and use that kubeconfig to communicate with the kubernetes api. Setting an annotation is easy proof...